### ISSUE TITLE
Heap memory build-up when MapContainer is removed from DOM

### ISSUE THREAD

**Issue by corrigancd** [PaulLeCam/react-leaflet#941]
## Bug report
Hello PaulLeCam and React-Leaflet, 

First of all, this is a great library and I appreciate all the effort that has gone into making it. At this point, I've exhausted all online documentation, stackoverflow (and equivalents), peer advice and my knowledge. I'd be interested in contributing some documentation on this issue, although I'd need some help understanding what is going on.   

- [x] All peer dependencies are installed: React, ReactDOM and Leaflet.
- [x] Using a supported version of React and ReactDOM (v17.0.0 minimum).
- [x] Using the supported version of Leaflet (v1.7.1 minimum) and its corresponding CSS file is loaded.
- [x] Using the [latest v3 version of React-Leaflet](https://github.com/PaulLeCam/react-leaflet/releases).
- [x] The issue has not already been reported.
- [x] Make sure you have followed the [quick start guide](http://leafletjs.com/examples/quick-start.html) for Leaflet.
- [x] Make sure you have fully read the [documentation](https://react-leaflet.js.org/docs/start-introduction) and that you understand the [limitations](https://react-leaflet.js.org/docs/start-introduction#limitations).

## Expected behavior

It makes sense that elements, removed from the DOM, are garbage collected over time in all of the scenarios outlined in the `Steps to reproduce` section.

## Actual behavior

Heap memory seems to build up over time when in a webpack/babel environment

## Steps to reproduce

The issue is that I am able to produce a memory leak when using React-Leaflet in a Webpack/Babel environment. However, this does not happen when I reproduce the issue in codepen. I'll describe each option I've tried below and describe the results. 

### Using create-react-app

I've produced a repo. with the minimum requirements to reproduce the issue - https://github.com/corrigancd/react-leaflet-query-minimised

In the repository linked above, I'm using the following versions of the required libraries (pasted here from the `package.json`) and I have the leaflet.css file loaded (although not required for my example):
```
    "leaflet": "^1.7.1",
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "react-leaflet": "^3.2.2",
```

The below is how to reproduce. I've included 2 scenarios. Scenario 1 has one React-Leaflet `MapContainer` and for scenario 2 I've replaced the `MapContainer` entirely with a div containing 1001 divs. Both scenarios use a `setInterval` in a `useEffect` hook to add/remove the `MapContainer`/`divs` from the DOM.

#### Steps to run the create-react-app example:
1. clone repository
2. make sure you are on master branch
3. Run `npm i`

##### Scenario 1
1. Run `npm start`
2. Hard reload Browser and navigate to `localhost:3000`
3. Open memory tab of Chrome dev. tools and observe that over the time, heap memory builds. 

It does take some time for memory to build. Using the console logs as a reference, I found heap memory had climbed from 3.8 mb (when count was 0) to 30.0 mb when the count, logged in the console, was 3280. 

##### Scenario 2
1. Checkout the branch `div` with `git checkout div`
2. Run `npm start`
3. Hard reload Browser and navigate to `localhost:3000`
4. Open memory tab of Chrome dev. tools and observe that over the time, heap memory fluctuates between 10 and 30 mb continuously over any number of iterations

**The only difference between `scenario 2` and `scenario 1` is that the `MapContainer` in Scenario 1 has been replaced with a div containing 1001 divs**

### Using standalone babel and webpack
No example here. Just did an extra bit of investigation in case there was something fishy with `create-react-app`. I was able to reproduce the results from scenario 1 in the `create-react-app` example with an environment using standalone webpack and babel, versions listed below:
```
    "webpack": "^4.46.0"
  },
  "devDependencies": {
    "@babel/core": "^7.9.6",
    "@babel/plugin-proposal-nullish-coalescing-operator": "^7.16.7",
    "@babel/preset-env": "^7.9.5",
    "@babel/preset-react": "^7.9.4",
    "babel-loader": "^8.1.0",
    "css-loader": "^3.5.2",
    "less": "^4.1.2",
    "less-loader": "^7.3.0",
    "postcss-loader": "^3.0.0",
    "style-loader": "^1.1.4",
    "ts-loader": "^7.0.0",
    "uglifyjs-webpack-plugin": "^2.2.0",
    "webpack-cli": "^3.3.11",
    "webpack-dev-server": "^3.10.3"
  },
```

### On Codepen
Using the CDN's and codepen (https://codepen.io/corrigancd/pen/RwLyKmB), all is ok. The map is removed from the DOM and the removed React-Leaflet objects are garbage collects as expected. Memory hovers in or around 11 mb. 

### Additional notes
- I've also tried running `leafletMap.remove()`, in the `create-react-app` environment and got the same heap memory build up;
- When I take heap snapshots and compare them, I noticed that it was the Leaflet objects still in memory;
- I was able to reproduce when on a production build genetrated from `create-react-app`
- When layers are added to the `MapContainer`, these are also not removed from the DOM. Using the described `setInterval` approach, I've got heap memory up to 1.3 gb, which crashed the browser.

**Comment by PaulLeCam** [comment]
If you can't reproduce the issue on CodePen it's likely coming from another part of your stack.
Please feel free to reopen if you find out the cause of the issue is actually in this library, but I won't investigate this myself unless it can be reproduced on CodePen.

**Comment by corrigancd** [comment]
Thanks for getting back, and I agree with isolating the problem as much as possible. I gave the codepen another go (https://codepen.io/corrigancd/pen/RwLyKmB) and have been able to reproduce. The heap memory starts at 3.7-3.8 mb:
![image](https://user-images.githubusercontent.com/36197976/148644026-9dcbc5f6-3375-4ce4-933d-71db0a0278dd.png)

And after letting it run, the memory grows to 30.2 mb after 1650 iterations (an iteration is the count logged to the console tab):
![image](https://user-images.githubusercontent.com/36197976/148645659-e89b2060-6a58-4479-a2e1-8ba3b10244c4.png)

**Comment by corrigancd** [comment]
> if you find out the cause of the issue is actually in this library,

@PaulLeCam, on trying to address this point, I've created a second CodePen. The Scenarios below are described in full in my openig post. 

In short, in Scenario 1, it is possible to see heap memory build up, while this does not happen in Scenario 2. The only difference between Scenario 1 and 2 is that the React Leaflet `MapContainer` implementation in Scenario 1 is replaced with a div containing 1001 divs in Scenario 2:

Scenario 1 - Map container: https://codepen.io/corrigancd/pen/RwLyKmB (screeshots above)
Scenario 2 - 1001 divs: https://codepen.io/corrigancd/pen/vYezXRd

The button to reopen the issue isn't available for me, do you think it may be worth reopening?

**Comment by PaulLeCam** [comment]
Thanks for the CodePen example, indeed it seems the created `Map` handlers are not being released when the `MapContainer` is unmounted.
I just released v3.2.4 that explicitly calls the `remove()` method on the map when unmounting, using [this fork of your CodePen](https://codepen.io/PaulLeCam/pen/JjraGzy) with the updated version seems to release memory as expected but please let me know if you still see the issue, thanks!

**Comment by corrigancd** [comment]
That's working nicely and also for my much larger create-react-app implementation.

Consider this one closed I think :+1: ,  really appreciate the quick turn around.

Thanks!
